File: packages/stow-s3-py/src/stow_s3/ready.py
"""Client half of the session readiness protocol.

The server writes one JSON object to a file descriptor. Parsing is strict on
purpose: an unknown protocol version or a missing field is a specific error, not
a partially populated session that fails later in a confusing place.

This is the same protocol the TypeScript client speaks. The two must agree on the
message shape, the protocol version, and the meaning of a zero limit.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

# The version of this message shape. A client rejects an unknown version rather
# than guessing, because a later version may add or change fields.
READY_PROTOCOL_VERSION = 1

PROTOCOL_MISMATCH = "protocol_mismatch"
INTERNAL = "internal"


class StowProtocolError(Exception):
    """A ready message this client cannot safely act on.

    The ``code`` mirrors the TypeScript client so a caller that moves between
    languages sees the same distinction: a version mismatch is something to
    report or upgrade around, while ``internal`` is a bug here.
    """

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True)
class StowCapabilities:
    """What a caller may rely on before issuing operations."""

    persistent: bool
    multipart: bool
    upstream: bool
    conditional_writes: bool
    presigned_urls: bool
    # A limit of 0 means the server reported no limit, which is distinct from a
    # limit that happens to be zero. It is not "unlimited" and not "unknown";
    # it is what the protocol says when nothing was reported.
    max_bytes: int
    max_objects: int
    max_request_bytes: int


@dataclass(frozen=True)
class StowReady:
    """The single readiness object the server writes."""

    protocol_version: int
    binary_version: str
    endpoint: str
    region: str
    access_key_id: str
    secret_access_key: str
    mode: str
    backend: str
    capabilities: StowCapabilities


def _require_string(source: dict[str, Any], field: str) -> str:
    value = source.get(field)
    if not isinstance(value, str) or not value:
        raise StowProtocolError(PROTOCOL_MISMATCH, f"ready message is missing {field}")
    return value


def _require_bool(source: dict[str, Any], field: str) -> bool:
    # isinstance alone is not enough to catch a number, because bool is a
    # subclass of int and a truthy 1 would otherwise pass as True.
    if not isinstance(source.get(field), bool):
        raise StowProtocolError(PROTOCOL_MISMATCH, f"ready message is missing {field}")
    return source[field]


def _require_number(source: dict[str, Any], field: str) -> float:
    value = source.get(field)
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise StowProtocolError(PROTOCOL_MISMATCH, f"ready message is missing {field}")
    if isinstance(value, float) and (value != value or value in (float("inf"), float("-inf"))):
        raise StowProtocolError(PROTOCOL_MISMATCH, f"ready message is missing {field}")
    return value


def parse_ready_message(line: str) -> StowReady:
    """Parse the readiness object.

    Raises :class:`StowProtocolError` on an unknown protocol version, a
    non-object payload, or any missing field.
    """
    try:
        decoded = json.loads(line)
    except json.JSONDecodeError as error:
        raise StowProtocolError(
            PROTOCOL_MISMATCH, f"ready message is not valid JSON: {error}"
        ) from error
    if not isinstance(decoded, dict):
        raise StowProtocolError(PROTOCOL_MISMATCH, "ready message is not a JSON object")

    version = _require_number(decoded, "protocolVersion")
    if version != READY_PROTOCOL_VERSION:
        raise StowProtocolError(
            PROTOCOL_MISMATCH,
            f"unsupported ready protocol version {version}; "
            f"this client speaks {READY_PROTOCOL_VERSION}",
        )

    capabilities = decoded.get("capabilities")
    if not isinstance(capabilities, dict):
        raise StowProtocolError(PROTOCOL_MISMATCH, "ready message is missing capabilities")

    return StowReady(
        protocol_version=int(version),
        binary_version=_require_string(decoded, "binaryVersion"),
        endpoint=_require_string(decoded, "endpoint"),
        region=_require_string(decoded, "region"),
        access_key_id=_require_string(decoded, "accessKeyId"),
        secret_access_key=_require_string(decoded, "secretAccessKey"),
        mode=_require_string(decoded, "mode"),
        backend=_require_string(decoded, "backend"),
        capabilities=StowCapabilities(
            persistent=_require_bool(capabilities, "persistent"),
            multipart=_require_bool(capabilities, "multipart"),
            upstream=_require_bool(capabilities, "upstream"),
            conditional_writes=_require_bool(capabilities, "conditionalWrites"),
            presigned_urls=_require_bool(capabilities, "presignedUrls"),
            max_bytes=int(_require_number(capabilities, "maxBytes")),
            max_objects=int(_require_number(capabilities, "maxObjects")),
            max_request_bytes=int(_require_number(capabilities, "maxRequestBytes")),
        ),
    )

File: packages/stow-s3/src/ready.ts
// Client half of the session readiness protocol.
//
// The server writes one JSON object to an inherited file descriptor. Parsing is
// strict on purpose: an unknown protocol version or a missing field is a
// specific error, not a partially populated session that fails later in a
// confusing place.
export const READY_PROTOCOL_VERSION = 1;

export interface StowReadyCapabilities {
  persistent: boolean;
  multipart: boolean;
  upstream: boolean;
  conditionalWrites: boolean;
  presignedUrls: boolean;
  maxBytes: number;
  maxObjects: number;
  maxRequestBytes: number;
}

export interface StowReady {
  protocolVersion: number;
  binaryVersion: string;
  endpoint: string;
  region: string;
  accessKeyId: string;
  secretAccessKey: string;
  mode: string;
  backend: string;
  capabilities: StowReadyCapabilities;
}

export class StowProtocolError extends Error {
  readonly code: "protocol_mismatch" | "internal";

  constructor(code: "protocol_mismatch" | "internal", message: string) {
    super(message);
    this.name = "StowProtocolError";
    this.code = code;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function requireString(source: Record<string, unknown>, field: string): string {
  const value = source[field];
  if (typeof value !== "string" || value.length === 0) {
    throw new StowProtocolError("protocol_mismatch", `ready message is missing ${field}`);
  }
  return value;
}

function requireBoolean(source: Record<string, unknown>, field: string): boolean {
  const value = source[field];
  if (typeof value !== "boolean") {
    throw new StowProtocolError("protocol_mismatch", `ready message is missing ${field}`);
  }
  return value;
}

function requireNumber(source: Record<string, unknown>, field: string): number {
  const value = source[field];
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new StowProtocolError("protocol_mismatch", `ready message is missing ${field}`);
  }
  return value;
}

/**
 * Parse the single readiness object. Throws StowProtocolError on an unknown
 * protocol version, on a non-object payload, or on any missing field.
 */
export function parseReadyMessage(line: string): StowReady {
  let decoded: unknown;
  try {
    decoded = JSON.parse(line);
  } catch (error) {
    throw new StowProtocolError(
      "protocol_mismatch",
      `ready message is not valid JSON: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
  if (!isRecord(decoded)) {
    throw new StowProtocolError("protocol_mismatch", "ready message is not a JSON object");
  }
  const version = requireNumber(decoded, "protocolVersion");
  if (version !== READY_PROTOCOL_VERSION) {
    throw new StowProtocolError(
      "protocol_mismatch",
      `unsupported ready protocol version ${version}; this client speaks ${READY_PROTOCOL_VERSION}`,
    );
  }
  const capabilities = decoded.capabilities;
  if (!isRecord(capabilities)) {
    throw new StowProtocolError("protocol_mismatch", "ready message is missing capabilities");
  }
  return {
    protocolVersion: version,
    binaryVersion: requireString(decoded, "binaryVersion"),
    endpoint: requireString(decoded, "endpoint"),
    region: requireString(decoded, "region"),
    accessKeyId: requireString(decoded, "accessKeyId"),
    secretAccessKey: requireString(decoded, "secretAccessKey"),
    mode: requireString(decoded, "mode"),
    backend: requireString(decoded, "backend"),
    capabilities: {
      persistent: requireBoolean(capabilities, "persistent"),
      multipart: requireBoolean(capabilities, "multipart"),
      upstream: requireBoolean(capabilities, "upstream"),
      conditionalWrites: requireBoolean(capabilities, "conditionalWrites"),
      presignedUrls: requireBoolean(capabilities, "presignedUrls"),
      maxBytes: requireNumber(capabilities, "maxBytes"),
      maxObjects: requireNumber(capabilities, "maxObjects"),
      maxRequestBytes: requireNumber(capabilities, "maxRequestBytes"),
    },
  };
}
